src/craftdef.cpp

```diff
@@ -975,37 +975,20 @@ class CCraftDefManager: public IWritableCraftDefManager
 			IGameDef *gamedef) const
 	{
 		std::vector<CraftDefinition*> recipes_list;
-		CraftInput input;
-		CraftOutput tmpout;
-		tmpout.item = "";
-		tmpout.time = 0;
 
-		for(std::vector<CraftDefinition*>::const_reverse_iterator
-				i = m_craft_definitions.rbegin();
-				i != m_craft_definitions.rend(); i++)
-		{
-			CraftDefinition *def = *i;
+		std::map<std::string, std::vector<CraftDefinition*> >::const_iterator vecIter =
+			m_output_to_craft_definition_map.find(output.item);
 
-			/*infostream<<"Checking "<<input.dump()<<std::endl
-					<<" against "<<def->dump()<<std::endl;*/
+		if (vecIter == m_output_to_craft_definition_map.end())
+			return recipes_list;
 
-			try {
-				tmpout = def->getOutput(input, gamedef);
-				if(tmpout.item.substr(0,output.item.length()) == output.item)
-				{
-					// Get output, then decrement input (if requested)
-					input = def->getInput(output, gamedef);
-					recipes_list.push_back(*i);
-				}
-			}
-			catch(SerializationError &e)
-			{
-				errorstream<<"getCraftResult: ERROR: "
-						<<"Serialization error in recipe "
-						<<def->dump()<<std::endl;
-				// then go on with the next craft definition
-			}
+		for(std::vector<CraftDefinition*>::const_reverse_iterator
+			i = vecIter->second.rbegin();
+			i != vecIter->second.rend(); i++)
+		{
+			recipes_list.push_back(*i);
 		}
+			
```

Comment:
Tables would break stuff. I'm ok with adding them, but I'd be already satisfied when there would be an additional function `get_actual_craft_recipes` which removes craft duplicate recipes from its output, using the last specified one, like `get_craft_recipe` [does](https://github.com/gregorycu/minetest/blob/Issue-2210/src/craftdef.cpp#L902-903). On the long term however, I think that tables are best.